Source/DafnyCore/Resolver/PreType/PreTypeConstraints.cs

```diff
@@ -650,8 +669,9 @@ public enum CommonConfirmationBag {
     /// If "super" is an ancestor of "sub", then return a list "L" of arguments for "super" such that
     /// "super<L>" is a supertype of "sub<subArguments>".
     /// Otherwise, return "null".
+    /// If "forAsOrIs" is "true", then allow "sub" to be replaced by an ancestor type of "sub" if "sub" is a newtype.
     /// </summary>
-    public List<PreType> /*?*/ GetTypeArgumentsForSuperType(TopLevelDecl super, TopLevelDecl sub, List<PreType> subArguments) {
+    public List<PreType> /*?*/ GetTypeArgumentsForSuperType(TopLevelDecl super, TopLevelDecl sub, List<PreType> subArguments, bool forAsOrIs) {
```

Comment:
Let's try to avoid specific naming AST constructs such as `as` and `is` in code that could otherwise be generic. Maybe something like `allowBaseTypeCast`.

You can replace the arguments `TopLevelDecl sub, List<PreType> subArguments` with a single `DPreType subType`. Then a further improvement is to add a field `ParentTypes` to `DPreType` that returns a list of DPreType, which have the same type parameters already filled in.

Reply:
I made these improvements, except that I added `ParentTypes` (which I named `ParentPreTypes`) in the current file rather than in class `DPreType`, because its implementation needs a `PreTypeResolver`. The change in signatures for `GetTypeArgumentsForSuperType` simplified all call sites.